### PR TITLE
Support Utf8View and BinaryView in substrait serialization.

### LINKED ISSUE #12118: Support substrait serialization  for `ScalarValue::Utf8View` and `ScalarValue::BinaryView`
### Is your feature request related to a problem or challenge?

Part of https://github.com/apache/datafusion/issues/11752

We are trying to enable DataFusion to use StringViewArray by default. If we do that it means `ScalarValue::Utf8View` and `ScalarValue::BinaryView` will be more likely to be used in plans

### Describe the solution you'd like

Thus we need to ensure `ScalarValue::Utf8View` and `ScalarValue::BinaryView`  can be serialized using datafusion substrait

### Describe alternatives you've considered

I recommend adding coverage for `ScalarValue::Utf8View` and `ScalarValue::BinaryView to the tests here

https://github.com/apache/datafusion/blob/b4069a65a9bb207370d382bdde93f1c98d69b9eb/datafusion/substrait/src/logical_plan/producer.rs#L2283-L2284

And then update the code to get the tests to pass

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wiedld**
## Which issue does this PR close?

Closes #12118 

## Rationale for this change

We have two new view data types, [Utf8View](https://docs.rs/arrow-schema/52.2.0/arrow_schema/enum.DataType.html#variant.Utf8View) and [BinaryView](https://docs.rs/arrow-schema/52.2.0/arrow_schema/enum.DataType.html#variant.BinaryView). Support in datafusion is part of [this epic](https://github.com/apache/datafusion/issues/11752), and this specific PR is about adding support for the (de-)serialization of logical and physical plans into the substrait format.

This PR adds new substrait variations on existing type classes. For example, there is a "string" substrait class which can have different variations representing different physical types (e.g. Utf8 vs LargeUtf8 vs Utf8View). If we serialize using string variation=2 (e.g. view physical type), then the deserialization of variation=2 will give us back the Utf8View. More background is [given here](https://github.com/apache/datafusion/issues/12118#issuecomment-2311396274).

## What changes are included in this PR?

* feat(12118): logical plan support for Utf8View (d7be771eb)
* feat(12118): physical plan support for Utf8View (b17ae25a7)
* feat(12118): logical plan support for BinaryView (f38085d4c)
* feat(12118): physical plan support for BinaryView (5c4ebec5c)

## Are these changes tested?

**Logical plan**: The Utf8View and BinaryView are covered in the logical plan roundtrip serialization tests.

**Physical plan**: However, the physical plan roundtrip serialization tests are [not yet implemented](https://github.com/apache/datafusion/blob/6ffb1f64a7b3f48c7693d2ecb34f72a2dca24acc/datafusion/substrait/tests/cases/roundtrip_physical_plan.rs#L82). There is an [ongoing epic](https://github.com/apache/datafusion/issues/5173) to finish the physical plan serialization. As such, I added code for the physical plan substrait handling of Utf8View and BinaryView (to avoid incurring more tech debt) -- but this code is not tested.

## Are there any user-facing changes?

No API contract change.
Removal of unimplemented errors if using these new datatypes in subtrait serialization.

### CODE REVIEW

File: datafusion/substrait/tests/cases/roundtrip_logical_plan.rs [hunk 1]
```diff
@@ -716,12 +716,29 @@ async fn all_type_literal() -> Result<()> {
...
+    let err = roundtrip_all_types(
+        "select * from data where
+            view_binary_col = arrow_cast('binary_view', 'BinaryView');",
+    )
+    .await;
+
+    assert!(
+        matches!(err, Err(e) if e.to_string().contains("Unsupported CAST from Utf8 to BinaryView"))
+    );
+    Ok(())
```
Comment (wiedld) [inline datafusion/substrait/tests/cases/roundtrip_logical_plan.rs:739]: Looks like we have a few other missing casts for BinaryView. Going to file a ticket and put up a PR in arrow-cast.

File: datafusion/substrait/tests/cases/roundtrip_logical_plan.rs [hunk 2]
```diff
@@ -716,12 +716,29 @@ async fn all_type_literal() -> Result<()> {
             date32_col = arrow_cast('2020-01-01', 'Date32') AND
             binary_col = arrow_cast('binary', 'Binary') AND
             large_binary_col = arrow_cast('large_binary', 'LargeBinary') AND
+            view_binary_col = arrow_cast(arrow_cast('binary_view', 'Binary'), 'BinaryView') AND
```
Comment (wiedld) [inline datafusion/substrait/tests/cases/roundtrip_logical_plan.rs:719]: See test `binaryview_type_literal_needs_casting_fix()` below, as for the reason behind the double casting.

File: datafusion/substrait/tests/cases/roundtrip_logical_plan.rs (hunk 1 above)
Comment (wiedld) [reply datafusion/substrait/tests/cases/roundtrip_logical_plan.rs:739]: I [see sqllogictests](https://github.com/apache/datafusion/blob/6ffb1f64a7b3f48c7693d2ecb34f72a2dca24acc/datafusion/sqllogictest/test_files/arrow_typeof.slt#L316) which demonstrate what is supported by arrow_cast. Then my follow ups will be: (a) make sqllogictests showing what is, and is not, supported of the new view types, and then (b) make the upstream arrow-rs changes (with some correctness guidance during code review).

Comment (wiedld) [reply datafusion/substrait/tests/cases/roundtrip_logical_plan.rs:739]: Sqllogictests added: https://github.com/apache/datafusion/pull/12200

File: datafusion/substrait/src/variation_const.rs [hunk 3]
```diff
@@ -52,6 +52,7 @@ pub const DATE_32_TYPE_VARIATION_REF: u32 = 0;
 pub const DATE_64_TYPE_VARIATION_REF: u32 = 1;
 pub const DEFAULT_CONTAINER_TYPE_VARIATION_REF: u32 = 0;
 pub const LARGE_CONTAINER_TYPE_VARIATION_REF: u32 = 1;
+pub const VIEW_CONTAINER_TYPE_VARIATION_REF: u32 = 2;
```
Comment (Blizzara) [inline datafusion/substrait/src/variation_const.rs:55]: FWIW, hardcoding the numbers isn't really the proper way to do type variations. (Rather we should add the variation as an extension and refer to the extension's id.) However, given this is already used for default vs large, I guess adding view makes sense - and they can all be migrated at once to the proper way someday.

Comment (wiedld) [reply datafusion/substrait/src/variation_const.rs:55]: Thank you. I'll craft up a follow up ticket later today, and link here (for future reference).

File: datafusion/substrait/src/logical_plan/producer.rs [hunk 4]
```diff
@@ -2351,8 +2371,10 @@ mod test {
         round_trip_type(DataType::Binary)?;
         round_trip_type(DataType::FixedSizeBinary(10))?;
         round_trip_type(DataType::LargeBinary)?;
+        round_trip_type(DataType::BinaryView)?;
```
Comment (alamb) [inline datafusion/substrait/src/logical_plan/producer.rs:2358]: 👍

File: datafusion/substrait/tests/cases/roundtrip_logical_plan.rs (hunk 1 above)
Comment (alamb) [reply datafusion/substrait/tests/cases/roundtrip_logical_plan.rs:739]: We have now updated to the latest arrow-rs so we'll have the correct code https://github.com/apache/datafusion/pull/12032

File: datafusion/substrait/tests/cases/roundtrip_logical_plan.rs (hunk 2 above)
Comment (alamb) [reply datafusion/substrait/tests/cases/roundtrip_logical_plan.rs:719]: I removed this workaround in a5bfedd0d

**Review by alamb — APPROVED** [review]
Thank you @wiedld  and @Blizzara  -- I think this PR is looking good. 

@wiedld  now that we have updated to the latest arrow-rs (https://github.com/apache/datafusion/pull/12032) I updated this PR to remove the workarounds. I figured I had it checked out anyways so I would just push the change

File: datafusion/substrait/tests/cases/roundtrip_logical_plan.rs (hunk 1 above)
Comment (wiedld) [reply datafusion/substrait/tests/cases/roundtrip_logical_plan.rs]: Sweet. I've removed the work-arounds. Thank you.

**Comment by alamb** [comment]
Thanks again @wiedld and @Blizzara

**Pull request merged**